src/rgw/rgw_op.cc

```diff
@@ -6132,6 +6137,45 @@ void RGWCompleteMultipart::execute(optional_yield y)
   }
 }
 
+bool RGWCompleteMultipart::check_previously_completed(const DoutPrefixProvider* dpp, const RGWMPObj& mpobj, const RGWMultiCompleteUpload* parts, rgw::sal::RGWStore* store)
+{
+  // TODO: replace with mpstat() in zipper
+  auto rados = dynamic_cast<rgw::sal::RGWRadosStore*>(store);
+  if (!rados) {
+    ldpp_dout(dpp, 0) << __func__ << "() ERROR: rados not avaliable" << dendl;
+    return false;
+  }
+  for (auto iter = parts->parts.begin() ; iter != parts->parts.end() ; ++iter) {
+    std::string oid = mpobj.get_part(iter->first);
```

Comment:
this is still doing a stat on each of the part objects

we should just stat the head object (which `RGWCompleteMultipart::execute()` calls `target_obj`) to get the xattrs for its manifest and etag

Reply:
will work on refactoring

Reply:
@cbodley refactored, performing stat on only the 1st part object to check the upload-id, then re-calculate the parts etag and compare with the etag from the xattrs